Migrate serialization format to standard PKCS#11 URIs compliant with RFC7512

dwmw2 opened the pull request:
This migrates the serialization format to conform to the PKCS#11 URI scheme as described in RFC7512.

The old form is still recognised for compatibility, but standard PKCS#11 URIs are now generated and accepted. Testing with OpenVPN now gives me the following output:

```
$ openvpn --show-pkcs11-ids /usr/lib64/pkcs11/gnome-keyring-pkcs11.so 

The following objects are available for use.
Each object shown below may be used as parameter to
--pkcs11-id option please remember to use single quote mark.

Certificate
       DN:             DC=com, DC=intel, DC=corp, DC=ger, OU=Workers, CN=Woodhouse, David, emailAddress=david.woodhouse@intel.com
       Serial:         1EB2ECCF00030058F375
       Serialized id:  pkcs11:model=1.0;manufacturer=Gnome%20Keyring;serial=1%3aUSER%3aDEFAULT;token=Gnome2%20Key%20Storage;id=%59%ae%17%70%af%e8%af%9f%5b%94%fb%c6%89%f6%f1%4c%11%5c%36%0e
```

... and the PKCS#11 URI thus generated is also usable with the `--pkcs11-id` option.

Commit:
Accept RFC7512-compliant PKCS#11 URIs as serialized token/certificate IDs

The old format is still accepted for compatibility.

Signed-off-by: David Woodhouse <David.Woodhouse@intel.com>

Commit:
Stop _pkcs11h_util_hexToBinary() checking for trailing NUL

We are going to want to use this for parsing %XX hex escapes in RFC7512
PKCS#11 URIs, where we cannot expect a trailing NUL. Since there's only
one existing caller at the moment, it's simple just to let the caller
have responsibility for that check.

Signed-off-by: David Woodhouse <David.Woodhouse@intel.com>

dwmw2 commented:
Now that the PKCS#11 URI format has been published as RFC7512 it would be very good to get this merged... should I collapse the above fixes into the original commits and resubmit? After this length of time there doesn't seem much point in keeping them incremental any more...

Commit:
Serialize to RFC7512-compliant PKCS#11 URIs

Signed-off-by: David Woodhouse <David.Woodhouse@intel.com>

nmav commented:
The pkcs11: URI format is a standard (RFC7512) since this April. It would be nice if pkcs11-helper would support that format.

dwmw2 commented:
Updated patch with minor fix from https://bugzilla.redhat.com/show_bug.cgi?id=1264645

dwmw2 commented:
Ping. It would be really good to get this merged...

alonbl commented:
Hi,
As I wrote, your implementation is much more complex than it should, it
won't get merged.
I need to rework this into something simpler, but this is extremely low
priority for me.
Thanks,
Alon

On 23 August 2016 at 16:47, dwmw2 <notifications@github.com> wrote:

> Ping. It would be really good to get this merged...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenSC/pkcs11-helper/pull/4#issuecomment-241736063>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABNIrSZPebF7EBGQmup94jOCvJftFw2Jks5qivncgaJpZM4DGws9>
> .
>

dwmw2 commented:
It could be massively simplified if it were to just use libp11-kit for the URI parsing. Other than that, I'm not sure how it could be simpler.

emmanuel-deloget commented:
Hi, 

RFC7512 is now two years old.  Could this patch (or any equivalent patch) be merged into pkcs11-helper ? 

Best regards, 

-- Emmanuel Deloget

dwmw2 commented:
Updated with bug fix from https://bugzilla.redhat.com/show_bug.cgi?id=1516474

celesteking commented:
Can we get this crap merged finally so openvpn actualy works with smartcards properly via NetworkManager? Thanks.

boomer41 commented:
Any updates on this one?

alonbl commented:
As I wrote, this patch is way to complex to achieve the desired task, and is partial as it supports only full pkcs11 url which as far as I understand is the opposite of reaching portability between applications that support the notation. I also got reports that the windows version of openvpn in which this patch is applied is failing to parse urls.

I am unsure how it is that important to network manager and am unsure why the openvpn management interface cannot be used to detect the available ids and apply id as opaque. And I keep reminding people that at daemon context openvpn should not have access to user's devices anyway, and the proper solution is to remove the entire pkcs11 support from the daemon and make sure the management interface is capable of delegating all private key operation to the user interface program, however, it seems like the openvpn project is on freeze.

dwmw2 commented:
Updated with fix for https://bugzilla.redhat.com/show_bug.cgi?id=1825496

> As I wrote, this patch is way to complex to achieve the desired task, 

I'm not quite sure how it could be simpler. I suppose it could take a library like `libp11-kit` as a dependency instead of doing the serialisation/deserialisation for itself?

> and is partial as it supports only full pkcs11 url which as far as I understand is the opposite of reaching portability between applications that support the notation. 

That doesn't seem to make much sense to me; can you explain what you mean? It will *output* the full URI when serialising, when listing all available objects. But obviously accepts only a partial URI (i.e. not redundantly specifying all attributes) when identifying an object. In that respect, it behaves like everything else does, and should.

> I also got reports that the windows version of openvpn in which this patch is applied is failing to parse urls.

A specific bug report would be very much appreciated, if there is misbehaviour.

> I am unsure how it is that important to network manager and am unsure why the openvpn management interface cannot be used to detect the available ids and apply id as opaque. 

There is work on displaying the available objects through a UI (like in Seahorse) and that widget will emit the RFC7512 ID identifying the chosen object. For that, we need to be using the standard.

> And I keep reminding people that at daemon context openvpn should not have access to user's devices anyway, and the proper solution is to remove the entire pkcs11 support from the daemon and make sure the management interface is capable of delegating all private key operation to the user interface program, however, it seems like the openvpn project is on freeze.

Yeah, it should do proxying from something in the user's session. But that's slightly orthogonal to the fact that we want to use a standard form to identify PKCS#11 objects.

becm commented:
OpenVPN uses an [out-of-date](https://github.com/OpenVPN/openvpn-build/pull/110) patch set which must fail on [16-byte](https://community.openvpn.net/openvpn/ticket/1075) token serials.

This was my [first problem](https://github.com/OpenVPN/openvpn-build/pull/172) as well, then [something else](https://bugzilla.redhat.com/show_bug.cgi?id=1825496) cropped up. Sadly neither seems to have made it in time for OpenVPN 2.4.9 release.

williamcroberts commented:
Literally without this patch, things are broken. Fedora patches their stuff with this. I just bumped into this, can we somehow get this merged or get a clear path forward?

becm commented:
The [menioned](#issuecomment-459986080) problems with `OpenVPN` on Windows should be [fixed](https://community.openvpn.net/openvpn/ticket/1075) by updated patch used in future releases.
Using the current `pkcs11-helper 1.26` will also address [EC support detection](OpenVPN/openvpn-build/issues/168) and [PSS padding](22) problems.

saper commented:
Seems like this change might break access to some HSMs 

https://bugzilla.redhat.com/show_bug.cgi?id=2298882 https://github.com/alonbl/gnupg-pkcs11-scd/issues/63

saper commented:
> your implementation is much more complex than it should

https://github.com/OpenSC/pkcs11-helper/pull/69 is my attempt to see if this can be done simpler. For now, this branch can only use URL encoding instead of `\x20`.

dwmw2 committed:
Accept empty token fields

It's a limitation of the core pkcs11-helper token matching code that we
need to specify *all* of model=, token=, manufacturer= and serial=. This
was true of the legacy serialization format, so it isn't a regression.

At least it *wouldn't* have been, if we it had distinguished between an
*explicit* "model=" parameter, and the model not being specified at all.
Thus https://bugzilla.redhat.com/show_bug.cgi?id=2298882

The requirement for all four token fields to be specified does mean that
applications using pkcs11-helper aren't *quite* as versatile and user
friendly as something which implements the full search algorithm shown in
§8 of http://david.woodhou.se/draft-woodhouse-cert-best-practice.html by
first searching for the specified certificate in all tokens without a login,
then only logging into the token in which the *certificate* was found, to
access the key. But that's OK, and something we can improve on later. It's
not a barrier to using the RFC7512 URI format in place of the legacy
serialization format.

dwmw2 commented:
Updated with fix for https://bugzilla.redhat.com/show_bug.cgi?id=2298882 (untested). Discussion in https://github.com/alonbl/gnupg-pkcs11-scd/issues/63

dwmw2 commented:
Does the fix work for you?

inorton-entrust commented:
> It continues to accept key identifiers in the old non-standard format; it now accepts both forms. It just _generates_ the RFC7512 standard form rather than the legacy form.
> 
> There waa simple bug in handling the RFC7512 form of identifier (as an input) when the `model=` field is empty, because of the way the underlying token selection works in pkcs11-helper. Does the fix work for you?

Thanks, I will build this PR branch and give it a try!

inorton-entrust commented:
Hi David,

Thanks for the email. I can see you’ve had a little trouble getting this merged upstream.   I’ve not been able to build it on fedora or rocky, do I need a newer autoconf ?

Ian

From: David Woodhouse ***@***.***>
Reply to: OpenSC/pkcs11-helper ***@***.***>
Date: Tuesday 19 November 2024 at 07:28
To: OpenSC/pkcs11-helper ***@***.***>
Cc: Ian Norton ***@***.***>, Comment ***@***.***>
Subject: [EXTERNAL] Re: [OpenSC/pkcs11-helper] Migrate serialization format to standard PKCS#11 URIs compliant with RFC7512 (#4)

Does the fix work for you? — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented. Message ID: <OpenSC/pkcs11-helper/pull/4/c2484887783@ github. com> ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍ ‍


Does the fix work for you?

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/OpenSC/pkcs11-helper/pull/4*issuecomment-2484887783__;Iw!!FJ-Y8qCqXTj2!eJsBJALNFY7GPQ4ak_ge4Hvq_SyWjZQGb7JpJt7RIKY0-5uz4CmgGX0Ss6reoYYmSidzl0QWIVtFSXK4ozHeKukZMg$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/BMVCN54FOKZ65TW5E4T4BYD2BLSAVAVCNFSM6AAAAABO7WAFM2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIOBUHA4DONZYGM__;!!FJ-Y8qCqXTj2!eJsBJALNFY7GPQ4ak_ge4Hvq_SyWjZQGb7JpJt7RIKY0-5uz4CmgGX0Ss6reoYYmSidzl0QWIVtFSXK4ozEUXUXJog$>.
You are receiving this because you commented.Message ID: ***@***.***>

Any email and files/attachments transmitted with it are intended solely for the use of the individual or entity to whom they are addressed. If this message has been sent to you in error, you must not copy, distribute or disclose of the information it contains. Please notify Entrust immediately and delete the message from your system.

inorton-entrust commented:
> It continues to accept key identifiers in the old non-standard format; it now accepts both forms. It just _generates_ the RFC7512 standard form rather than the legacy form.
> 
> There waa simple bug in handling the RFC7512 form of identifier (as an input) when the `model=` field is empty, because of the way the underlying token selection works in pkcs11-helper. Does the fix work for you?

Hi David,  I've been able to build this branch and test it.  Things fail early because of how the "serial number" of the smartcard is reported.

This is computed from the sha1 hash of the "serialized" token info that this branch adds.  So now gpg thinks it has a different card in the reader.

I'll experiment and see what I can come up with.

dwmw2 commented:
For a moment I thought you meant the `serial=` part of the URI, from the `serialNumber` field of `CK_TOKEN_INFO`. But that isn't it, is it? You mean a serial number generated internally by GPG to match the token (or the key)?

inorton-entrust commented:
> For a moment I thought you meant the `serial=` part of the URI, from the `serialNumber` field of `CK_TOKEN_INFO`. But that isn't it, is it? You mean a serial number generated internally by GPG to match the token (or the key)?

The gnupg-pkcs11-scd app needs to return a value for it's `SERIALNO` command to the gpg agent. `gnupg-pcs11-scd` gets this value by calling `pkcs11h_token_serializeTokenId()` and then feeds that string into sha1, it uses the first 8 hex chars of this sha1 hash to form the "card serial" .  This substring is embedded into the middle of the `SERIALNO` output.

With the unmodified pkcs11-helper you get:

```
$ echo SERIALNO | gnupg-pkcs11-scd --server
S SERIALNO D27600012401115031312CF4EEE61111 0
```
With the uri-scheme output of `pkcs11h_token_serializeTokenId()` the middile `2CF4EEE6` substring changes when given the same token.

saper commented:
Changing serial numbers is not good. GnuPG needs stable numbers, since they are stored in the placeholders of the private key. Frankly, I am no longer sure if this change is worth the hassle.

inorton-entrust commented:
> Changing serial numbers is not good. GnuPG needs stable numbers, since they are stored in the placeholders of the private key. Frankly, I am no longer sure if this change is worth the hassle.

If this were early days I'd agree it needs more planning, but Redhat based distros have been shipping with this as a patch for nearly a decade. I do have a PR that can allow inter-operation with both styles controlled by an env var

saper commented:
What is the problem currently with OpenVPN? Do you have any pointers?
